### ISSUE TITLE
Runtime increasing with more cores (homogenous and mostly empty boxes)

### ISSUE THREAD

**Issue by JimmyHolloway** [SmileiPIC/Smilei#10]
Hi SMILEI developers,

I investigated how SMILEI's runtime scales with number of cores it is running on for a small plasma in a vacuum and for a homogenous plasma filling the simulaiton box. I also varied the number of patches coverign the 2D boxes.

The runtime generally increases with increasing number of cores and the number of patches does not seem to aleviate this.

The below plot shows the runtime as a function of number of cores and number of patches, with the runtime being denoted by the size of the data point. This is for a 'cold core' simulation (temperature too low for Debye length to be resolved).
![coldcore](https://user-images.githubusercontent.com/26409910/27638779-385bd0f4-5c0c-11e7-8206-2c38d8d3d0dd.png)

For a hot core csimulation (Debye length resolved):
![hotcore](https://user-images.githubusercontent.com/26409910/27638858-7262195c-5c0c-11e7-8596-da654f4dc928.png)

For a hot homogenous plasma:
![hothomo](https://user-images.githubusercontent.com/26409910/27638862-7bb59a74-5c0c-11e7-9079-c23ecf7b7dc2.png)

A hot homogenous plasma for a large number of cores with the various run time metrics given by the log file plotted:
![hothomogenouscorescan](https://user-images.githubusercontent.com/26409910/27638885-8fad2b0a-5c0c-11e7-9a2d-f06e073eaf31.png)

Notes:
The log file indicates I am utilizing the many cores.
The input decks and log files can by found at the bottom of this post.
The input decks are missing a line or two that defines the number of patchs (this is because a script build the final decks from the input deck provided).
The number of patches in the above plots is in base 2. i.e. 1,2,4,8,16 etc.
I am running on SCARF lexicon-2 at the rutherford appleton laboratory.
I have renamed .log and .py extensions to .txt so I can upload to this post.
I can provide the MATLAB .fig files that generated the .png files if you wish to pull the numbers out / rotate freely.
I can provide further details upon request.

All the best,
Jimmy.

[coldcore-222.txt](https://github.com/SmileiPIC/Smilei/files/1108811/coldcore-222.txt)
[coldcoreinput.txt](https://github.com/SmileiPIC/Smilei/files/1108812/coldcoreinput.txt)
[hotcore-222.txt](https://github.com/SmileiPIC/Smilei/files/1108813/hotcore-222.txt)
[hotcoreinput.txt](https://github.com/SmileiPIC/Smilei/files/1108814/hotcoreinput.txt)
[hothomo-222.txt](https://github.com/SmileiPIC/Smilei/files/1108815/hothomo-222.txt)
[hothomoinput.txt](https://github.com/SmileiPIC/Smilei/files/1108816/hothomoinput.txt)
[hothomogenouscorescan256cores.txt](https://github.com/SmileiPIC/Smilei/files/1108817/hothomogenouscorescan256cores.txt)
[hothomogenouscorescaninput.txt](https://github.com/SmileiPIC/Smilei/files/1108818/hothomogenouscorescaninput.txt)

**Comment by mccoys** [comment]
Hi,

I have a few comments, that might help you investigate the causes of this problem, unless you have already tried all of that. Let me try anyway.

Concerning simulations with a cold or hot core, there is always going to be one MPI that takes all the work to do. To reduce this problem you may use load balancing:

- Have you activated [MPI load balancing](http://www.maisondelasimulation.fr/smilei/namelist.html#load-balancing) ? It could help move patches around MPI processes.

- Have you tried making use of OpenMP ? Your simulations seem to use only 1 openMP thread per MPI. You should investigate on the architecture of your machines. Typically, you must have as many MPI processes as **nodes** and as many threads-per-MPI as you have cores-per-node. This situation can vary.

- To have efficient local (OpenMP) and global (MPI) load balancing, you must ensure you have **more** patches than the total number of threads. You should also make sure that your patches are smaller than the core itself, in order to share the particles between several patches.

Concerning the hot homogeneous simulation, this is a bit puzzling. From your outputs, it seems that most of the simulation time is taken in the synchronisation of particles. However, with a decently low temperature, the time for synchronisations should be low compared to the time for computing particles. If the temperature is very high, there might be too many communications because the particles travel fast, causing the synchronisation lag. This still looks surprising. Are you sure the simulation stays homogeneous over time?

**Comment by jderouillat** [comment]
Hi,
I ran the `hothomogenouscorescaninput.py` simulation on our cluster (each node have 2 Intel Sandy Bridge processor of 8 cores) for 1 to 256 MPI process (1 thread OpenMP).  
The test is homogeneous so I didn't oversplit the domain in patches, I used 1 patch per MPI and no dynamic load balancing.  

I split the domain to be as square as possible, and the only change in the namelist is the suppression of the definition of patchesx and patchesy that I modifief in the command line:  
```
$ mpirun -np 1 ./smilei  hothomogenouscorescaninput.py Main.number_of_patches=[1,1]  2>&1 |tee smilei_001_MPI_001_OMP.log
$ mpirun -np 2 ./smilei  hothomogenouscorescaninput.py Main.number_of_patches=[2,1]  2>&1 |tee smilei_002_MPI_001_OMP.log
$ mpirun -np 4 ./smilei  hothomogenouscorescaninput.py Main.number_of_patches=[2,2]  2>&1 |tee smilei_004_MPI_001_OMP.log
$ mpirun -np 8 ./smilei  hothomogenouscorescaninput.py Main.number_of_patches=[4,2]  2>&1 |tee smilei_008_MPI_001_OMP.log
$ mpirun -np 16 ./smilei  hothomogenouscorescaninput.py Main.number_of_patches=[4,4]  2>&1 |tee smilei_016_MPI_001_OMP.log
$ mpirun -np 32 -ppn 16 ./smilei  hothomogenouscorescaninput.py Main.number_of_patches=[8,4]  2>&1 |tee smilei_032_MPI_001_OMP.log
$ mpirun -np 64 -ppn 16 ./smilei  hothomogenouscorescaninput.py Main.number_of_patches=[8,8]  2>&1 |tee smilei_064_MPI_001_OMP.log
$ mpirun -np 128 -ppn 16 ./smilei  hothomogenouscorescaninput.py Main.number_of_patches=[16,8]  2>&1 |tee smilei_128_MPI_001_OMP.log
$ mpirun -np 256 -ppn 16 ./smilei  hothomogenouscorescaninput.py Main.number_of_patches=[16,16] 2>&1 |tee smilei_256_MPI_001_OMP_clrw.log
```

For my point of view results are good considering the total number of particles.  
On the 256 MPI process simulation, the number of particles per MPI process considering the 2 species is 1600 :  
```
$ grep "Time in time loop" smilei*
smilei_001_MPI_001_OMP.log: Time in time loop :	74.567	99.843% coverage
smilei_002_MPI_001_OMP.log: Time in time loop :	35.904	99.835% coverage
smilei_004_MPI_001_OMP.log: Time in time loop :	18.314	99.785% coverage
smilei_008_MPI_001_OMP.log: Time in time loop :	9.738	99.722% coverage
smilei_016_MPI_001_OMP.log: Time in time loop :	5.421	99.466% coverage
smilei_032_MPI_001_OMP.log: Time in time loop :	2.813	99.247% coverage
smilei_064_MPI_001_OMP.log: Time in time loop :	1.653	98.879% coverage
smilei_128_MPI_001_OMP.log: Time in time loop :	1.114	97.852% coverage
smilei_256_MPI_001_OMP.log: Time in time loop :	0.895	97.031% coverage
```

**Comment by jderouillat** [comment]
I looked at `hotcoreinput.py` too, for this beginning of simulation it will be very hard to get an honest scaling on more than 4 compute units.  
At initialization, the plasma is distributed on a circle which have a radius of 1.5.  
The cell_length is 0.2, so with a minimum patch size of 6 cells, a patch will almost cover a quarter of the plasma, 4 patches cover almost the plasma. At the end of this part of the simulation, particles haven't move enough to consider a better scaling.
Below the scalability for 1 to 4 MPI / patches :  
```
$ grep "Time in time" smilei*.log
smilei_001_MPI_001_OMP.log: Time in time loop :	16.202	98.869% coverage
smilei_002_MPI_001_OMP.log: Time in time loop :	8.825	98.801% coverage
smilei_004_MPI_001_OMP.log: Time in time loop :	4.533	98.629% coverage
```

**Comment by JimmyHolloway** [comment]
Hi Mccoys, Jderouillat,

I took your advice and made use of open MP and load balancing and investigated the temperature's effect on runtime.

The high temperatures:
I ran a cold-homogenous simulation set to compare with the hot-homogenous set. The cold-homogenous scaled sensibly. The log file told me that the code spent a lot less time synchronised partciles and fields in the cold-homo simulations. I am convinced the problem was just the really high temperature of the hot-homo simulations.
![coldhomo](https://user-images.githubusercontent.com/26409910/28016082-b7d2aa6c-656a-11e7-9163-1145e6a1680a.png)

Open MP:
I performed the same simulation set as above but with open MP used properly and got good computational savings:
![coldhomoopm](https://user-images.githubusercontent.com/26409910/28016128-f35fc2ea-656a-11e7-9048-5cbdc57a8590.png)

I am still testing load balancing but it looks like it gives further computaitonal saves with the 'core' simulations.

Thank you both for helping me with this!
Best,
Jimmy.
